### PR TITLE
Convert layers to support eager mode

### PR DESCRIPTION
**Pull request by peastman**
This is very preliminary: I've only converted one nontrivial layer so far.  I'm posting it now so you can see the approach I'm taking.

TensorGraph will also need a lot of changes to support eager mode.  I might save that for a separate PR to keep the changes more manageable.

### CODE REVIEW

**Comment by peastman** [comment]
A few more layers converted, and lots more tests written.  A few notes on this:

I'm replacing all uses of `tf.contrib.layers` functions with the corresponding `tf.layers` classes.  The contrib versions don't have an object oriented interface, so you can't separate the creation of the variables from computing the output.

I changed the API for `Highway` so that the number of output channels is a constructor argument.  Without that, it can't create the variables in the constructor.  This also allows it to compute the output shape, which it wasn't doing before, and makes it more consistent with other layers.

Does anyone understand how `CombineMeanStd` is supposed to work?  According to the docs, the second input tensor specifies the standard deviation of the noise for each element.  But there's also a constructor argument called `noise_epsilon`, and the docs also claim *that* specifies the standard deviation of the noise!  As for what is actually happening, the two get combined in a way I don't understand:

```python
    sample_noise = tf.random_normal(
        mean_parent.get_shape(), 0, self.noise_epsilon, dtype=tf.float32)
    ...
    out_tensor = mean_parent + tf.exp(std_parent * 0.5) * sample_noise
```

**Comment by rbharath** [comment]
@peastman This looks amazing! Took a look through the test suite and it looks like an amazing peek into what the DeepChem API might look like in a couple releases. I see it becoming much easier to build complex models :)

+1 for the shift to `tf.layers` away from `tf.contrib`. It's good for us to get close to the standard TF implementation path. Standardizing `Highway` sounds good to me as well. Unfortunately, don't know about CombineMeanStd though.

**Comment by peastman** [comment]
I'm a bit stuck on `GraphConv`.  When `create_tensor()` is called, is looks at the shape of the first input and uses it to determine the number of input channels:

```python
in_channels = inputs[0].get_shape()[-1].value
```

The shape of the variables it creates depends on that.  Do I need to add a new argument to the constructor?  Is there some other way to figure out the required shape in the constructor?

**Comment by peastman** [comment]
`BatchNormalization` has the same problem.  Not to be confused with `BatchNorm`.  Why do we have two different classes for batch normalization?

Anyway, I'm thinking maybe I should redesign this.  It turns out Tensorflow layers don't create their variables in the constructor.  They wait until the first time they're called so they know the size of the inputs (and they assume those sizes will always be the same).

**Comment by rbharath** [comment]
We should get rid of one of the two BatchNorm classes. There's been some confusion about the two of them already.

I like the idea of creating variables upon first invocation. That seems to match intuitive understanding of how these layers work and allows us to sidestep the issues with GraphConv and other complex models.

**Comment by peastman** [comment]
This is ready to merge.  The only failure is an unrelated stochastic one.

**Comment by rbharath** [comment]
@peastman This looks great! I agree with the strategy of handling the graph layers in a second pass. Will open up an issue to coordinate.

I did a review pass and things look good to me. However, this is a large PR, so I think we might want to do a 2 person review for safety. 

@lilleswing Would you mind reviewing this PR on your end as well?

**Comment by rbharath** [comment]
Started #1186 to coordinate adding tests for the graph layers.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+
+
+class TestLayersEager(test_util.TensorFlowTestCase):
+  """
+  Test that layers function in eager mode.
+  """
+
+  def test_conv_1d(self):
+    """Test invoking Conv1D in eager mode."""
+    with context.eager_mode():
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_layers_eager.py]: nit: can only have one level of indentation with multiple with syntax.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+      with tfe.IsolateTest():
+        width = 5
+        in_channels = 2
+        filters = 3
+        kernel_size = 2
+        batch_size = 10
+        input = np.random.rand(batch_size, width, in_channels).astype(
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_layers_eager.py]: nit: input is a python keyword

File: deepchem/models/tensorgraph/tests/test_layers_eager.py [hunk 3]
```diff
@@ -0,0 +1,980 @@
...
+            np.float32)
+        layer = layers.Conv1D(filters, kernel_size)
+        result = layer(input)
+        self.assertEqual(result.shape[0], batch_size)
+        self.assertEqual(result.shape[2], filters)
+        assert len(layer.variables) == 2
+
+        # Creating a second layer should produce different results, since it has
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_layers_eager.py]: Maybe a dumb question but how would the syntax work if you wanted to reuse the weights? for a new "input"

File: deepchem/models/tensorgraph/tests/test_layers_eager.py [hunk 4]
```diff
@@ -0,0 +1,980 @@
+import deepchem as dc
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_layers_eager.py]: Our test coverage is going to go through the roof :fire:

File: deepchem/models/tensorgraph/layers.py [hunk 5]
```diff
@@ -365,6 +376,8 @@ def shared(self, in_layers):
     return copy
 
   def _get_scope_name(self):
+    if tfe.in_eager_mode():
+      return None
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: I'm a bit confused by this -- I imagine sharing variables in eager is much easier but how does it work?

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: I'm not sure what you mean?  The code works correctly.  The indentation is set by yapf.  Both "with" statements are required for it to work correctly.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 2 above)
Comment (peastman) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: I don't think so?  It's not listed as one at https://www.programiz.com/python-programming/keywords-identifier.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 3 above)
Comment (peastman) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: Just invoke it again with a different input: `result2 = layer(input2)`.

File: deepchem/models/tensorgraph/layers.py (hunk 5 above)
Comment (peastman) [reply deepchem/models/tensorgraph/layers.py]: Currently it's not supported.  For the purposes of this PR (invoking layers directly), there's no need to share weights.  You just invoke the same layer again on different inputs.  That's how Tensorflow layers are designed to be used.  In the followup PR to get TensorGraph working in eager mode, I do want to try to support sharing variables.  I haven't figured out how I'll implement that yet.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 1 above)
Comment (lilleswing) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: It is a python language construct that I didn't know about until recently.  I'm not saying we have to use it or that we even should use it I just thought I could share.

https://stackoverflow.com/questions/893333/multiple-variables-in-python-with-statement
```python
with context.eager_mode() as _, tfe.IsolateTest() as _:
  # All The Code
```

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 2 above)
Comment (lilleswing) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: It is a python built-in function not a keyword.  using it as a variable overwrites the variable from the built-in function to the variable but it does keep the weird highlighting by text editors as if it were a function as seen here.

We don't have to change it just a comment.

File: deepchem/models/tensorgraph/tests/test_layers_eager.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/tensorgraph/tests/test_layers_eager.py]: Cool, I didn't know about that feature.

**Comment by lilleswing** [comment]
LGTM!
